File: client_sdk/java/the-client/src/vip/dummy/texasholdem/WebSocketClient.java
package vip.dummy.texasholdem;

import java.io.IOException;

import javax.websocket.ClientEndpoint;
import javax.websocket.OnClose;
import javax.websocket.OnError;
import javax.websocket.OnMessage;
import javax.websocket.OnOpen;
import javax.websocket.Session;

import com.google.gson.Gson;
import vip.dummy.texasholdem.bean.Credential;
import vip.dummy.texasholdem.indication.*;
import vip.dummy.texasholdem.message.JoinMessage;
import vip.dummy.texasholdem.message.data.JoinData;
import vip.dummy.texasholdem.param.SampleConfigurator;
import vip.dummy.texasholdem.param.SampleDecoder;
import vip.dummy.texasholdem.param.SampleEncoder;
import vip.dummy.texasholdem.playerai.PlayerAI;

/**
 * Created by the-engine team
 * 2017-12-17
 *
 * Texas Hold'em AI client web socket interface
 */
@ClientEndpoint(
		configurator = SampleConfigurator.class,
		decoders={SampleDecoder.class},
		encoders={SampleEncoder.class},
		subprotocols={})
public class WebSocketClient {

	private Session session;
	private Credential credential;
	private String ticket;

    private PlayerAI playerAI;

	private static final String NEW_PEER = "__new_peer";
	private static final String NEW_ROUND = "__new_round";
    private static final String START_RELOAD = "__start_reload";
	private static final String DEAL = "__deal";
	private static final String ACTION = "__action";
	private static final String BET = "__bet";
	private static final String SHOW_ACTION = "__show_action";
	private static final String ROUND_END = "__round_end";
	private static final String GAME_OVER = "__game_over";

	WebSocketClient(Credential credential, String ticket) {
	    this.credential = credential;
	    this.ticket = ticket;
	    this.playerAI = new PlayerAI(this);
    }

	@OnOpen
	public void onOpen(Session session) {
		System.out.println("Client WebSocket is opening...");
		this.session = session;

		// send join message here with your name
		JoinData joinData = new JoinData(credential.getPhoneNumber(), credential.getPassword(), ticket);
		JoinMessage joinMessage = new JoinMessage(joinData);
		String joinString = new Gson().toJson(joinMessage);
		System.out.println("joinString = " + joinString);
		send(joinString);
	}

	@OnMessage
	public void onMessage(String message) {
		System.out.println("received message: " + message);
		// parse message for player AI
        Indication indication = new Gson().fromJson(message, Indication.class);
        switch(indication.getEventName()) {
            case NEW_PEER:
                /*
                 * __new_peer
                 */
                NewPeerIndication newPeerIndication = new Gson().fromJson(message, NewPeerIndication.class);
                playerAI.onNewPeer(newPeerIndication);
                break;

            case NEW_ROUND:
                /*
                 * __new_round
                 */
                NewRoundIndication newRoundIndication = new Gson().fromJson(message, NewRoundIndication.class);
                playerAI.onNewRound(newRoundIndication);
                break;

            case START_RELOAD:
                /*
                 * __start_reload
                 */
                StartReloadIndication startReloadIndication = new Gson().fromJson(message, StartReloadIndication.class);
                playerAI.onStartReload(startReloadIndication);
                break;

            case DEAL:
                /*
                 * __deal
                 */
                DealIndication dealIndication = new Gson().fromJson(message, DealIndication.class);
                playerAI.onDeal(dealIndication);
                break;

            case ACTION:
                /*
                 * __action
                 */
                ActionIndication actionIndication = new Gson().fromJson(message, ActionIndication.class);
                playerAI.onAction(actionIndication);
                break;

            case BET:
                /*
                 * __bet
                 */
                BetIndication betIndication = new Gson().fromJson(message, BetIndication.class);
                playerAI.onBet(betIndication);
                break;

            case SHOW_ACTION:
                /*
                 * __show_action
                 */
                ShowActionIndication showActionIndication = new Gson().fromJson(message, ShowActionIndication.class);
                playerAI.onShowAction(showActionIndication);
                break;

            case ROUND_END:
                /*
                 * __round_end
                 */
                RoundEndIndication roundEndIndication = new Gson().fromJson(message, RoundEndIndication.class);
                playerAI.onRoundEnd(roundEndIndication);
                break;

            case GAME_OVER:
                /*
                 * __game_over
                 */
                GameOverIndication gameOverIndication = new Gson().fromJson(message, GameOverIndication.class);
                playerAI.onGameOver(gameOverIndication);
                break;

            default:
                break;
        }
    }

	@OnClose
	public void onClose() {
        System.out.println("Web socket closed");
    }

    @OnError
    public void onError(Session session, Throwable t) {
        t.printStackTrace();
    }

	public void send(Object message) {
		this.session.getAsyncRemote().sendObject(message);
	}
	
	public void close() throws IOException {
		if (this.session.isOpen()) {
			this.session.close();
		}
	}
}

File: client_sdk/python/the-client/src/vip/dummy/texasholdem/WebSocketClient.py
# -*- coding: utf-8 -*-
import sys
import json
from argparse import Namespace

sys.path.append('./bean')
sys.path.append('./message')
sys.path.append('./message/data')
sys.path.append('./indication')
sys.path.append('./indication/data')
sys.path.append('./playerai')
from Credential import Credential
from PlayerAI import PlayerAI
from IndicationCallbacks import IndicationCallbacks
from ReloadMessage import ReloadMessage
from ReloadData import ReloadData
from ActionMessage import ActionMessage
from ActionData import ActionData
from JoinMessage import JoinMessage
from JoinData import JoinData
from PlayerAI import PlayerAI
from Indication import Indication
from ActionData import ActionData
from BetData import BetData
from DealData import DealData
from GameOverData import GameOverData
from NewRoundData import NewRoundData
from RoundEndData import RoundEndData
from ShowActionData import ShowActionData
from StartReloadData import StartReloadData
from ActionIndication import ActionIndication
from BetIndication import BetIndication
from DealIndication import DealIndication
from GameOverIndication import GameOverIndication
from NewPeerIndication import NewPeerIndication
from NewRoundIndication import NewRoundIndication
from RoundEndIndication import RoundEndIndication
from ShowActionIndication import ShowActionIndication
from StartReloadIndication import StartReloadIndication

NEW_PEER = "__new_peer"
NEW_ROUND = "__new_round"
START_RELOAD = "__start_reload"
DEAL = "__deal"
ACTION = "__action"
BET = "__bet"
SHOW_ACTION = "__show_action"
ROUND_END = "__round_end"
GAME_OVER = "__game_over"


class WebSocketClient(IndicationCallbacks):
    def __init__(self, credential):
        self.credential = credential
        self.ticket = credential.ticket
        self.gameName = credential.gameName
        self.playerAI = PlayerAI(self)

    def convert_to_dict(self, obj):
        dict = {}
        dict.update(obj.__dict__)
        return dict

    def onOpen(self, session):
        print ("Client WebSocket is opening...")
        self.session = session
        joinData = JoinData(self.credential.phoneNumber, self.credential.password, self.ticket,self.gameName)
        joinMessage = JoinMessage(self.convert_to_dict(joinData))
        joinString = self.convert_to_dict(joinMessage)
        self.send(joinString)

    def onMessage(self, message):
        try:
            message = json.loads(message, object_hook=lambda d: Namespace(**d))  # 转成对象才能用属性访问
            eventName = message.eventName
            data = message.data
            if eventName == NEW_PEER:
                newPeerIndication = NewPeerIndication(data)
                self.playerAI.onNewPeer(newPeerIndication)
            elif eventName == NEW_ROUND:
                newRoundData = NewRoundData(data.table, data.players)
                newRoundIndication = NewRoundIndication(newRoundData)
                self.playerAI.onNewRound(newRoundIndication)
            elif eventName == START_RELOAD:
                startReloadData = StartReloadData(data.tableNumber, data.players)
                startReloadIndication = StartReloadIndication(startReloadData)
                self.playerAI.onNewRound(startReloadIndication)
            elif eventName == DEAL:
                dealData = DealData(data.table, data.players)
                dealIndication = DealIndication(dealData)
                self.playerAI.onNewRound(dealIndication)
            elif eventName == ACTION:
                actionData = ActionData(data.tableNumber, data.self, data.game)
                actionIndication = ActionIndication(actionData)
                self.playerAI.onNewRound(actionIndication)
            elif eventName == BET:
                betData = BetData(data.tableNumber, data.self, data.game)
                betIndication = BetIndication(betData)
                self.playerAI.onNewRound(betIndication)
            elif eventName == SHOW_ACTION:
                showActionData = ShowActionData(data.action, data.table, data.players)
                showActionIndication = ShowActionIndication(showActionData)
                self.playerAI.onNewRound(showActionIndication)
            elif eventName == ROUND_END:
                roundEndData = RoundEndData(data.table, data.players)
                roundEndIndication = RoundEndIndication(roundEndData)
                self.playerAI.onNewRound(roundEndIndication)
            elif eventName == GAME_OVER:
                gameOverData = GameOverData(data.table, data.players, data.winners)
                gameOverIndication = GameOverIndication(gameOverData)
                self.playerAI.onNewRound(gameOverIndication)

        except Exception, e:
            print e.message

    def send(self, message):
        self.session.send(json.dumps(message))
